### ISSUE TITLE
Rendering issues with structures

### ISSUE THREAD

**Issue by solonovamax** [wisp-forest/lavender#2]
There is an issue with how structures are rendered, due to the depth test being enabled.

Patchouli fixes this by having their own custom `VertexConsumerProvider.Immediate` and `RenderLayer` implementations: https://github.com/VazkiiMods/Patchouli/blob/1.20.x/Xplat/src/main/java/vazkii/patchouli/client/handler/MultiblockVisualizationHandler.java#L340-L389

Implementing similar code myself fixes this:
<details>
<summary>Code Diff</summary>

```diff
--- old.StructureOverlayRenderer.java	2024-05-30 23:40:17.382540109 -0400
+++ StructureOverlayRenderer.java	2024-05-30 23:40:55.880873515 -0400
@@ -16,15 +16,19 @@
 import io.wispforest.owo.ui.event.WindowResizeCallback;
 import io.wispforest.owo.ui.hud.Hud;
 import io.wispforest.owo.ui.util.Delta;
+import it.unimi.dsi.fastutil.objects.Object2ObjectLinkedOpenHashMap;
 import net.fabricmc.fabric.api.client.rendering.v1.WorldRenderEvents;
 import net.fabricmc.fabric.api.event.player.UseBlockCallback;
 import net.minecraft.block.BlockState;
 import net.minecraft.client.MinecraftClient;
 import net.minecraft.client.gl.Framebuffer;
 import net.minecraft.client.gl.SimpleFramebuffer;
+import net.minecraft.client.render.BufferBuilder;
 import net.minecraft.client.render.LightmapTextureManager;
 import net.minecraft.client.render.OverlayTexture;
 import net.minecraft.client.render.OverlayVertexConsumer;
+import net.minecraft.client.render.RenderLayer;
+import net.minecraft.client.render.VertexConsumer;
 import net.minecraft.client.render.VertexConsumerProvider;
 import net.minecraft.client.render.VertexConsumerProvider.Immediate;
 import net.minecraft.client.render.model.ModelLoader;
@@ -44,8 +48,11 @@
 import org.apache.commons.lang3.mutable.MutableBoolean;
 import org.jetbrains.annotations.Nullable;
 import org.lwjgl.opengl.GL30C;
+import vazkii.patchouli.api.PatchouliAPI;
+import vazkii.patchouli.mixin.client.AccessorMultiBufferSource;

 import java.util.HashMap;
+import java.util.IdentityHashMap;
 import java.util.Map;
 import java.util.function.Supplier;

@@ -140,6 +147,7 @@

             var client = MinecraftClient.getInstance();
             var effectConsumers = client.getBufferBuilders().getEffectVertexConsumers();
+            var buffers = initBuffers(client.getBufferBuilders().getEntityVertexConsumers());
             var testPos = new BlockPos.Mutable();

             hudComponent.<FlowLayout>configure(layout -> {
@@ -179,7 +187,7 @@
                             }

                             if (entry.visibleLayer != -1 && pos.getY() != entry.visibleLayer) return;
-                            renderOverlayBlock(matrices, context.consumers(), pos, predicate, entry.rotation);
+                            renderOverlayBlock(matrices, buffers, pos, predicate, entry.rotation);

                         }, entry.rotation);

@@ -228,13 +236,14 @@
                         if (!client.player.isSneaking()) targetPos = targetPos.offset(target.getSide());

                         matrices.translate(targetPos.getX(), targetPos.getY(), targetPos.getZ());
-                        structure.forEachPredicate((pos, predicate) -> renderOverlayBlock(matrices, context.consumers(), pos, predicate, PENDING_OVERLAY.rotation), PENDING_OVERLAY.rotation);
+                        structure.forEachPredicate((pos, predicate) -> renderOverlayBlock(matrices, buffers, pos, predicate, PENDING_OVERLAY.rotation), PENDING_OVERLAY.rotation);
                     }
                 } else {
                     PENDING_OVERLAY = null;
                 }
             }

+            // buffers.
             matrices.pop();

             var framebuffer = FRAMEBUFFER.get();
@@ -244,7 +252,9 @@
             GL30C.glBindFramebuffer(GL30C.GL_READ_FRAMEBUFFER, client.getFramebuffer().fbo);
             GL30C.glBlitFramebuffer(0, 0, framebuffer.textureWidth, framebuffer.textureHeight, 0, 0, client.getFramebuffer().textureWidth, client.getFramebuffer().textureHeight, GL30C.GL_DEPTH_BUFFER_BIT, GL30C.GL_NEAREST);

-            if (context.consumers() instanceof VertexConsumerProvider.Immediate immediate) immediate.draw();
+            // if (context.consumers() instanceof VertexConsumerProvider.Immediate immediate)
+            //     immediate.draw();
+            buffers.draw();
             effectConsumers.draw();
             client.getFramebuffer().beginWrite(false);

@@ -329,4 +339,61 @@
         }
     }

+    private static VertexConsumerProvider.Immediate initBuffers(VertexConsumerProvider.Immediate original) {
+		BufferBuilder fallback = ((AccessorMultiBufferSource) original).getFallbackBuffer();
+		Map<RenderLayer, BufferBuilder> layerBuffers = ((AccessorMultiBufferSource) original).getFixedBuffers();
+		Map<RenderLayer, BufferBuilder> remapped = new Object2ObjectLinkedOpenHashMap<>();
+		for (Map.Entry<RenderLayer, BufferBuilder> e : layerBuffers.entrySet()) {
+			remapped.put(GhostRenderLayer.remap(e.getKey()), e.getValue());
+		}
+		return new GhostVertexConsumerProvider(fallback, remapped);
+	}
+
+    private static class GhostVertexConsumerProvider extends VertexConsumerProvider.Immediate {
+
+        protected GhostVertexConsumerProvider(BufferBuilder fallbackBuffer, Map<RenderLayer, BufferBuilder> layerBuffers) {
+            super(fallbackBuffer, layerBuffers);
+        }
+
+        @Override
+        public VertexConsumer getBuffer(RenderLayer renderLayer) {
+            return super.getBuffer(GhostRenderLayer.remap(renderLayer));
+        }
+    }
+
+    private static class GhostRenderLayer extends RenderLayer {
+        private static final Map<RenderLayer, RenderLayer> remappedTypes = new IdentityHashMap<>();
+
+        private GhostRenderLayer(RenderLayer original) {
+            super(
+                    String.format("%s_%s_ghost", original.toString(), PatchouliAPI.MOD_ID),
+                    original.getVertexFormat(),
+                    original.getDrawMode(),
+                    original.getExpectedBufferSize(),
+                    original.hasCrumbling(),
+                    true,
+                    () -> {
+                        original.startDrawing();
+
+                        RenderSystem.disableDepthTest();
+                        RenderSystem.enableBlend();
+                        RenderSystem.setShaderColor(1, 1, 1, 0.75f);
+                    },
+                    () -> {
+                        RenderSystem.setShaderColor(1, 1, 1, 1);
+                        RenderSystem.disableBlend();
+                        RenderSystem.enableDepthTest();
+
+                        original.endDrawing();
+                    }
+            );
+        }
+
+        public static RenderLayer remap(RenderLayer in) {
+            if (in instanceof GhostRenderLayer)
+                return in;
+            else
+                return GhostRenderLayer.remappedTypes.computeIfAbsent(in, GhostRenderLayer::new);
+        }
+    }
 }
```

</details

however, I'm also not a rendering person so I'm unsure of what like 75% of that code does, and it very well could be redundant or could be able to be avoided.

**Comment by gliscowo** [comment]
Patchouli's implementation of structure preview rendering is absolutely cringetastic and not code that will ever find its way into Lavender.

Anyways though (and this really, *really* should've been part of your initial report), what is the actual issue you're referring to? I'm of course open to fixing Lavender's implementation if there are indeed some issues with it we have never come across in our testing

**Comment by solonovamax** [comment]
oh yeah I should probably provide screenshots

imma do that later

**Comment by solonovamax** [comment]
### Lavender
- Funny depth issues
- back faces of blocks are rendered when they shouldn't be.
  Note: only happens at *some* angles.
- Can't see if a block behind other blocks is missing

![image](https://github.com/wisp-forest/lavender/assets/46940694/99eabb5c-a0c3-4edf-91f9-84d37a86bf4d)
![image](https://github.com/wisp-forest/lavender/assets/46940694/30916bd8-0f52-4316-9e6b-b9ff887e9b12)
![image](https://github.com/wisp-forest/lavender/assets/46940694/cfda702b-ebf2-4073-960f-be096b8745f4)
![image](https://github.com/wisp-forest/lavender/assets/46940694/c518c4d9-efc2-418d-a6e6-44c4028ceff7)

### Patchouli

- Does not render block backfaces
- The completion bar looks, arguably, nicer (though this is entirely subjective and not at all related to the issue lol)
- Can see missing blocks behind other blocks

![image](https://github.com/wisp-forest/lavender/assets/46940694/025fb721-d2c7-40eb-a51a-b4314c111d72)
![image](https://github.com/wisp-forest/lavender/assets/46940694/51388ba0-c02a-4a25-a822-9ed8d0a6af07)
![image](https://github.com/wisp-forest/lavender/assets/46940694/ae7aeabf-b14d-48f3-b92c-a89838c25cc8)

**Comment by solonovamax** [comment]
Here's another screenshot of an issue with the rendering, which I just encountered:
![2024-06-12_16 32 27](https://github.com/wisp-forest/lavender/assets/46940694/f1545ee8-1a02-4bbc-94eb-860133fc08b0)

**Comment by gliscowo** [comment]
Do you have any reproduction cases for these weird (seemingly depth-related?) rendering problems? Neither me nor anyone else I know of has ever experienced anything like them. This makes sense if you consider how Lavender's structure rendering works - the structure is rendered completely normally (read, without any weird translucency hacks) into a offscreen buffer which is then drawn back into the main buffer with 50% opacity. This way, no hacks are required and it should (in theory at least) work for any and all blocks which use custom rendering. 

Now, this does of course have the issue where blocks that are hidden behind others don't render at all - however, this is an intentional decision. I personally believe (and I have found that most everyone else agrees) that while Patchouli's rendering makes it *possible* to see blocks behind others it's not exactly easy or great to use. This is precisely why Lavender structures can be previewed layer-by-layer. That way, there is no messy translucency like with Patchouli *and* all blocks are easily discernible.

**Comment by solonovamax** [comment]
> Do you have any reproduction cases for these weird (seemingly depth-related?) rendering problems? Neither me nor anyone else I know of has ever experienced anything like them. This makes sense if you consider how Lavender's structure rendering works - the structure is rendered completely normally (read, without any weird translucency hacks) into a offscreen buffer which is then drawn back into the main buffer with 50% opacity. This way, no hacks are required and it should (in theory at least) work for any and all blocks which use custom rendering.
> 
> Now, this does of course have the issue where blocks that are hidden behind others don't render at all - however, this is an intentional decision. I personally believe (and I have found that most everyone else agrees) that while Patchouli's rendering makes it _possible_ to see blocks behind others it's not exactly easy or great to use. This is precisely why Lavender structures can be previewed layer-by-layer. That way, there is no messy translucency like with Patchouli _and_ all blocks are easily discernible.

I believe that it may actually have been caused my another mod that was being loaded in my dev env, possibly sodium 0.5.1. I am currently trying to see if that is the case.

**Comment by solonovamax** [comment]
It seems the issue has something to do with the "Fabulous!" graphics setting.

- 1.20.1 with sodium 0.5.8:
  Shows the back faces of blocks at certain angles, as shown above.
- 1.20.4 with sodium 0.5.8:
  Shows the back faces of blocks at certain angles, as shown above.
- 1.20.1 without sodium:
  Does not render the preview what so ever.
- 1.20.4 without sodium:
  Does not render the preview what so ever.


I believe it's due to this code in `RenderLayers` (which is invoked by `vertexConsumers.getBuffer(state, false)` in `BlockRenderManager.renderBlockAsEntity()`):
```java
if (!MinecraftClient.isFabulousGraphicsOrBetter()) {
    return TexturedRenderLayers.getEntityTranslucentCull();
} else {
    return direct ? TexturedRenderLayers.getEntityTranslucentCull() : TexturedRenderLayers.getItemEntityTranslucentCull();
}
```

Here is a patch that fixes it:

```patch
From ccd5992a2cff5a86d3f8bfb016ce9151710326ff Mon Sep 17 00:00:00 2001
From: solonovamax <solonovamax@12oclockpoint.com>
Date: Wed, 12 Jun 2024 18:55:44 -0400
Subject: [PATCH] Fix rendering issue when graphics setting is set to fabulous

Copy out BlockRenderManager.renderBlockAsEntity() and hardcode the layer to be TexturedRenderLayers.getEntityTranslucentCull().

Signed-off-by: solonovamax <solonovamax@12oclockpoint.com>
---
 .../client/StructureOverlayRenderer.java      | 49 ++++++++++++++++++-
 .../access/BlockRenderManagerAccessor.java    | 20 ++++++++
 src/main/resources/lavender.mixins.json       |  1 +
 3 files changed, 69 insertions(+), 1 deletion(-)
 create mode 100644 src/main/java/io/wispforest/lavender/mixin/access/BlockRenderManagerAccessor.java

diff --git a/src/main/java/io/wispforest/lavender/client/StructureOverlayRenderer.java b/src/main/java/io/wispforest/lavender/client/StructureOverlayRenderer.java
index 8415230..95518ce 100644
--- a/src/main/java/io/wispforest/lavender/client/StructureOverlayRenderer.java
+++ b/src/main/java/io/wispforest/lavender/client/StructureOverlayRenderer.java
@@ -3,6 +3,7 @@ package io.wispforest.lavender.client;
 import com.google.common.base.Suppliers;
 import com.mojang.blaze3d.systems.RenderSystem;
 import io.wispforest.lavender.Lavender;
+import io.wispforest.lavender.mixin.access.BlockRenderManagerAccessor;
 import io.wispforest.lavender.structure.BlockStatePredicate;
 import io.wispforest.lavender.structure.LavenderStructures;
 import io.wispforest.lavender.structure.StructureTemplate;
@@ -15,15 +16,21 @@ import io.wispforest.owo.ui.hud.Hud;
 import io.wispforest.owo.ui.util.Delta;
 import net.fabricmc.fabric.api.client.rendering.v1.WorldRenderEvents;
 import net.fabricmc.fabric.api.event.player.UseBlockCallback;
+import net.minecraft.block.BlockRenderType;
+import net.minecraft.block.BlockState;
 import net.minecraft.client.MinecraftClient;
 import net.minecraft.client.gl.Framebuffer;
 import net.minecraft.client.gl.SimpleFramebuffer;
 import net.minecraft.client.render.LightmapTextureManager;
 import net.minecraft.client.render.OverlayTexture;
 import net.minecraft.client.render.OverlayVertexConsumer;
+import net.minecraft.client.render.TexturedRenderLayers;
 import net.minecraft.client.render.VertexConsumerProvider;
+import net.minecraft.client.render.model.BakedModel;
 import net.minecraft.client.render.model.ModelLoader;
+import net.minecraft.client.render.model.json.ModelTransformationMode;
 import net.minecraft.client.util.math.MatrixStack;
+import net.minecraft.item.ItemStack;
 import net.minecraft.sound.SoundCategory;
 import net.minecraft.sound.SoundEvents;
 import net.minecraft.text.Text;
@@ -292,7 +299,7 @@ public class StructureOverlayRenderer {
         matrices.scale(1.0001f, 1.0001f, 1.0001f);
         matrices.translate(-.5, -.5, -.5);
 
-        MinecraftClient.getInstance().getBlockRenderManager().renderBlockAsEntity(
+        renderBlockAsEntity(
                 block.preview().rotate(rotation),
                 matrices,
                 consumers,
@@ -302,6 +309,46 @@ public class StructureOverlayRenderer {
         matrices.pop();
     }
 
+    private static void renderBlockAsEntity(BlockState state, MatrixStack matrices, VertexConsumerProvider vertexConsumers, int light, int overlay) {
+        var renderManager = MinecraftClient.getInstance().getBlockRenderManager();
+        var blockModelRenderer = ((BlockRenderManagerAccessor) renderManager).getBlockModelRenderer();
+        var builtinModelItemRenderer = ((BlockRenderManagerAccessor) renderManager).getBuiltinModelItemRenderer();
+        BlockRenderType blockRenderType = state.getRenderType();
+        if (blockRenderType != BlockRenderType.INVISIBLE) {
+            switch (blockRenderType) {
+                case MODEL:
+                    BakedModel bakedModel = renderManager.getModel(state);
+                    int i = ((BlockRenderManagerAccessor) renderManager).getBlockColors().getColor(state, null, null, 0);
+                    float r = (i >> 16 & 0xFF) / 255.0F;
+                    float g = (i >> 8 & 0xFF) / 255.0F;
+                    float b = (i & 0xFF) / 255.0F;
+                    blockModelRenderer
+                        .render(
+                            matrices.peek(),
+                            vertexConsumers.getBuffer(TexturedRenderLayers.getEntityTranslucentCull()),
+                            state,
+                            bakedModel,
+                            r,
+                            g,
+                            b,
+                            light,
+                            overlay
+                        );
+                    break;
+                case ENTITYBLOCK_ANIMATED:
+                    builtinModelItemRenderer
+                        .render(
+                            new ItemStack(state.getBlock()),
+                            ModelTransformationMode.NONE,
+                            matrices,
+                            vertexConsumers,
+                            light,
+                            overlay
+                        );
+            }
+        }
+    }
+
     private static class OverlayEntry {
 
         public final Identifier structureId;
diff --git a/src/main/java/io/wispforest/lavender/mixin/access/BlockRenderManagerAccessor.java b/src/main/java/io/wispforest/lavender/mixin/access/BlockRenderManagerAccessor.java
new file mode 100644
index 0000000..ec26a90
--- /dev/null
+++ b/src/main/java/io/wispforest/lavender/mixin/access/BlockRenderManagerAccessor.java
@@ -0,0 +1,20 @@
+package io.wispforest.lavender.mixin.access;
+
+import net.minecraft.client.color.block.BlockColors;
+import net.minecraft.client.render.block.BlockModelRenderer;
+import net.minecraft.client.render.block.BlockRenderManager;
+import net.minecraft.client.render.item.BuiltinModelItemRenderer;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.gen.Accessor;
+
+@Mixin(BlockRenderManager.class)
+public interface BlockRenderManagerAccessor {
+    @Accessor
+    BlockColors getBlockColors();
+
+    @Accessor
+    BuiltinModelItemRenderer getBuiltinModelItemRenderer();
+
+    @Accessor
+    BlockModelRenderer getBlockModelRenderer();
+}
diff --git a/src/main/resources/lavender.mixins.json b/src/main/resources/lavender.mixins.json
index ab04798..7c2c049 100644
--- a/src/main/resources/lavender.mixins.json
+++ b/src/main/resources/lavender.mixins.json
@@ -9,6 +9,7 @@
     "access.SmithingTrimRecipeAccessor"
   ],
   "client": [
+    "access.BlockRenderManagerAccessor",
     "ClientAdvancementManagerAccessor",
     "CreativeInventoryScreenMixin",
     "DrawContextMixin",
-- 
2.45.2
```

**Comment by solonovamax** [comment]
I noticed you added some changes to fix the rendering in fabulous mode. unsure what the difference is between that and what my code here is doing, tbh, but I'm not an expert at funny rendering stuff

**Comment by gliscowo** [comment]
The primary difference is that my patch is significantly more general and works with blocks that use different renderlayers among other things. Anyways, since this has now been fixed I'll go ahead and close the issue

Cheers

**Comment by solonovamax** [comment]
could I maybe pretty please get it also merged into the 1.20 branch, bc rn my mod uses that?

bc if so, that would be epic
